src/Libraries/Microsoft.Extensions.Http.Telemetry/Logging/Internal/HttpClientLogger.cs

```diff
@@ -0,0 +1,234 @@
+            return null;
+        }
+    }
+
+    public async ValueTask LogRequestStopAsync(
+        object? context,
+        HttpRequestMessage request,
+        HttpResponseMessage response,
+        TimeSpan elapsed,
+        CancellationToken cancellationToken = default)
+            => await LogResponseAsync(context, request, response, null, elapsed, cancellationToken).ConfigureAwait(false);
+
+    public async ValueTask LogRequestFailedAsync(
+        object? context,
+        HttpRequestMessage request,
+        HttpResponseMessage? response,
+        Exception exception,
+        TimeSpan elapsed,
+        CancellationToken cancellationToken = default)
+            => await LogResponseAsync(context, request, response, exception, elapsed, cancellationToken).ConfigureAwait(false);
+
+    public object? LogRequestStart(HttpRequestMessage request)
+    {
+        throw new NotSupportedException();
+    }
+
+    public void LogRequestStop(object? context, HttpRequestMessage request, HttpResponseMessage response, TimeSpan elapsed)
+    {
+        throw new NotSupportedException();
+    }
+
+    public void LogRequestFailed(object? context, HttpRequestMessage request, HttpResponseMessage? response, Exception exception, TimeSpan elapsed)
+    {
+        throw new NotSupportedException();
+    }
+
+    private static LogLevel GetLogLevel(LogRecord logRecord)
+    {
+        const int HttpErrorsRangeStart = 400;
+        const int HttpErrorsRangeEnd = 599;
+        int statusCode = logRecord.StatusCode!.Value;
+
+        if (statusCode >= HttpErrorsRangeStart && statusCode <= HttpErrorsRangeEnd)
+        {
+            return LogLevel.Error;
+        }
+
+        return LogLevel.Information;
+    }
+
+    [SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "The logger shouldn't throw")]
+    private async ValueTask LogResponseAsync(
+        object? context,
+        HttpRequestMessage request,
+        HttpResponseMessage? response,
+        Exception? exception,
+        TimeSpan elapsed,
+        CancellationToken cancellationToken)
+    {
+        if (context is not LogRecord logRecord)
+        {
+            // TODO: we need to decide - log an error or try to load the log record from the context?
+            return;
+        }
+
+        LogMethodHelper? propertyBag = null;
+        List<KeyValuePair<string, string>>? responseHeadersBuffer = null;
+        try
+        {
+            if (response is not null)
+            {
+                if (_logResponseHeaders)
+                {
+                    responseHeadersBuffer = _headersPool.Get();
+                }
+
+                await _httpRequestReader.ReadResponseAsync(logRecord, response, responseHeadersBuffer, cancellationToken).ConfigureAwait(false);
+            }
+
+            propertyBag = LogMethodHelper.GetHelper();
+            FillLogRecord(logRecord, propertyBag, in elapsed, request, response, exception);
```

Comment:
We don't use `TimeProvider` here since `elapsed` is provided out-of-the-box, but we can consider ignoring it and using the old logic